### PR TITLE
Base Image and Video blocks on identical media components & reduce state instability

### PR DESCRIPTION
**Pull request by nathggns**
## 🌟 What is the purpose of this PR?

Successor to https://github.com/hashintel/hash/pull/347 

This fixes at least three bugs in the image and video blocks, while improving code readability. It also bases both on identical Media blocks, though these are not yet shared due to an issue with importing these.

## ⚠️ Known issues

- Image and Video blocks are using multiple pieces of state that need to be updated in tandem. To avoid multiple renders, I'm using an unofficial batch API. I'm doing this because:

 - A) The blocks are already doing this
 - B) Doing this simplifies some of the "revert to initial state" logic
 - C) We want to move to a reducer here anyway
 - D) React 18 batches by default

## 🚫 Blocked by

N/A

## 🐾 Next steps

- De-duplicate Media component between Image and Video blocks https://app.asana.com/0/1201095311341924/1201906715110980 _(internal)_

## 🔍 What does this change?

- All different pieces of state are now in individual states. This makes it easier read, but proliferates `unstable_batchedUpdates`. We already plan to move to a reducer to fix this
- New `useDefaultState` hook which makes it easy to have state-set-by-prop-as-default-but-reset-when-props-change
- UI parts of the image block have been compentized. This should ensure we react to props changes better.
- When you mouse up when changing width, ensure the event handler which sets the width is then removed, so it doesn't get called on every subsequent click
- Ensure the latest on width change function gets called when changing width, to prevent reverting to previous state
- Both Image and Video blocks are now based on a duplicated but identical `Media` block. We want this to be shared. 

These changes should fix a symptom where changing blocks doesn't properly reset the state.

### 📜 Does this require a change to the docs?

No

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1200211978612931/1201882912244908/f) (_internal_)

## 🛡 What tests cover this?

No

## ❓ How to test this?

Most of the bugs are present in the hub but not when tested in isolation

1.  Change the width of the block
2. Change the caption using the UI
3. Change the props passed by the hub
4. Change caption again
5. Ensure everything updates as expected

### CODE REVIEW

File: packages/blocks/image/src/components/Media.tsx [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+// https://www.typescriptlang.org/docs/handbook/release-notes/overview.html#recursive-conditional-types
+type Awaited<T> = T extends PromiseLike<infer U> ? Awaited<U> : T;
+type FileType = Awaited<ReturnType<BlockProtocolUploadFileFunction>>;
+type UpdateEntitiesMediaAction = {
+  initialCaption?: string;
+  initialWidth?: number;
+  file?: FileType;
+};
+
+const useDefaultState = <T extends any>(
```
Comment (CiaranMn) [inline packages/blocks/image/src/components/Media.tsx:31]: To capture the fact it won't work for primitive values (could have a comment instead)
```suggestion
const useDefaultState = <
  T extends number | string | boolean | null | undefined,
>(
```

File: packages/blocks/image/src/components/Media.tsx [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+    linkGroups,
+    path,
+    sourceEntityId,
+  });
+
+  if (!matchingLinkGroup) {
+    return null;
+  }
+
+  const destinationEntityId = matchingLinkGroup.links[0].destinationEntityId;
```
Comment (CiaranMn) [inline packages/blocks/image/src/components/Media.tsx:95]: Appreciate you only moved this, but might be worth handling checking this index access now so we don't have to when we introduce the tsconfig flag

File: packages/blocks/video/src/components/Media.tsx [hunk 3]
```diff
@@ -0,0 +1,366 @@
+import {
```
Comment (CiaranMn) [inline packages/blocks/video/src/components/Media.tsx:1]: Can we write somewhere in this file and other duplicated files that these are duplicated and `@todo` to share them? and advise people that changes in one place need to be made in the other until we share them

**Review by CiaranMn — DISMISSED** [review]
Thanks Nate - a few non-blocking comments

File: packages/blocks/video/src/components/Media.tsx (hunk 3 above)
Comment (nathggns) [reply packages/blocks/video/src/components/Media.tsx:9]: 👍

File: packages/blocks/image/src/components/Media.tsx (hunk 2 above)
Comment (nathggns) [reply packages/blocks/image/src/components/Media.tsx]: 👍

File: packages/blocks/image/src/components/Media.tsx (hunk 1 above)
Comment (nathggns) [reply packages/blocks/image/src/components/Media.tsx]: Good idea – done!

**Pull request merged**